### PR TITLE
Implemented epsilon-lexicase selection in selection.cljc and calculated epsilons once per generation in gp.cljc

### PR DESCRIPTION
**Pull request by dndang23**

### CODE REVIEW

File: src/propeller/gp.cljc [hunk 1]
```diff
@@ -13,6 +13,38 @@
...
+
+(defn epsilon-list
+  [pop]
+  (let [error-list (map :errors pop)
+        length (count (:errors (first pop)))]
+    (loop [epsilons [] i 0]
+      (if (= i length)
+        epsilons
+        (recur (conj epsilons (median-absolute-deviation (map #(nth % i) error-list))) (inc i))))))
+
```
Comment (ryanboldi) [inline src/propeller/gp.cljc:48]: I would not define these in gp.cljc

I think add all the math stuff to src/propeller/tools/math.cljc

I would also added epsilon-list to a file in selection, and then import it.

File: src/propeller/gp.cljc [hunk 2]
```diff
@@ -50,7 +82,11 @@
                                  (mapper
                                    (partial error-function argmap (:training-data argmap))
                                    population))
-          best-individual (first evaluated-pop)]
+          best-individual (first evaluated-pop)
+          argmap (if (= (:parent-selection argmap) :epsilon-lexicase)
+                           (assoc argmap :epsilons (epsilon-list evaluated-pop))
+                           argmap)]
+      (prn argmap)
```
Comment (ryanboldi) [inline src/propeller/gp.cljc]: I would avoid printing the entire argmap.

**Review by ryanboldi — CHANGES_REQUESTED** [review]
Looks good to me! I think some small organizational tweaks are needed but besides that it looks good.

Also, it would be nice to write some tests for the epsilon-generation functions when you move them to selection.

**Review by dndang23 — COMMENTED** [review]
Hi Ryan, could you review the changes I made based on your comments?

**Review by ryanboldi — APPROVED** [review]
Looks good now. Good work!

**Comment by lspector** [comment]
Thanks @dndang23 and @ryanboldi for your work on this! I've also patched up the formatting.

**Pull request merged**